File: app/src/main/java/com/fstyle/structure_android/data/source/remote_api/middleware/RxErrorHandlingCallAdapterFactory.java
package com.fstyle.structure_android.data.source.remote_api.middleware;

import android.text.TextUtils;
import android.util.Log;
import com.fstyle.structure_android.data.source.remote_api.error.BaseException;
import com.fstyle.structure_android.data.source.remote_api.response.ErrorResponse;
import com.google.gson.Gson;
import io.reactivex.Completable;
import io.reactivex.CompletableSource;
import io.reactivex.Flowable;
import io.reactivex.Maybe;
import io.reactivex.MaybeSource;
import io.reactivex.Observable;
import io.reactivex.ObservableSource;
import io.reactivex.Single;
import io.reactivex.SingleSource;
import io.reactivex.functions.Function;
import java.io.IOException;
import java.lang.annotation.Annotation;
import java.lang.reflect.ParameterizedType;
import java.lang.reflect.Type;
import org.reactivestreams.Publisher;
import retrofit2.Call;
import retrofit2.CallAdapter;
import retrofit2.HttpException;
import retrofit2.Response;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory;

/**
 * ErrorHandling:
 * http://bytes.babbel.com/en/articles/2016-03-16-retrofit2-rxjava-error-handling.html
 * This class only for Call in retrofit 2
 */
public final class RxErrorHandlingCallAdapterFactory extends CallAdapter.Factory {

    private static final String TAG = RxErrorHandlingCallAdapterFactory.class.getName();

    private final RxJava2CallAdapterFactory original;

    private RxErrorHandlingCallAdapterFactory() {
        original = RxJava2CallAdapterFactory.create();
    }

    public static CallAdapter.Factory create() {
        return new RxErrorHandlingCallAdapterFactory();
    }

    @Override
    public CallAdapter<?, ?> get(Type returnType, Annotation[] annotations, Retrofit retrofit) {
        return new RxCallAdapterWrapper(returnType,
                original.get(returnType, annotations, retrofit));
    }

    /**
     * RxCallAdapterWrapper
     */
    class RxCallAdapterWrapper<R> implements CallAdapter<R, Object> {
        private final Type returnType;
        private final CallAdapter<R, Object> wrapped;

        RxCallAdapterWrapper(Type type, CallAdapter<R, Object> wrapped) {
            returnType = type;
            this.wrapped = wrapped;
        }

        @Override
        public Type responseType() {
            return wrapped.responseType();
        }

        @Override
        public Object adapt(Call<R> call) {
            Class<?> rawType = getRawType(returnType);

            boolean isFlowable = rawType == Flowable.class;
            boolean isSingle = rawType == Single.class;
            boolean isMaybe = rawType == Maybe.class;
            boolean isCompletable = rawType == Completable.class;
            if (rawType != Observable.class && !isFlowable && !isSingle && !isMaybe) {
                return null;
            }
            if (!(returnType instanceof ParameterizedType)) {
                String name = isFlowable ? "Flowable"
                        : isSingle ? "Single" : isMaybe ? "Maybe" : "Observable";
                throw new IllegalStateException(name
                        + " return type must be parameterized"
                        + " as "
                        + name
                        + "<Foo> or "
                        + name
                        + "<? extends Foo>");
            }

            if (isFlowable) {
                return ((Flowable) wrapped.adapt(call)).onErrorResumeNext(
                        new Function<Throwable, Publisher>() {
                            @Override
                            public Publisher apply(Throwable throwable) throws Exception {
                                return Flowable.error(convertToBaseException(throwable));
                            }
                        });
            }
            if (isSingle) {
                return ((Single) wrapped.adapt(call)).onErrorResumeNext(
                        new Function<Throwable, SingleSource>() {
                            @Override
                            public SingleSource apply(Throwable throwable) throws Exception {
                                return Single.error(convertToBaseException(throwable));
                            }
                        });
            }
            if (isMaybe) {
                return ((Maybe) wrapped.adapt(call)).onErrorResumeNext(
                        new Function<Throwable, MaybeSource>() {
                            @Override
                            public MaybeSource apply(Throwable throwable) throws Exception {
                                return Maybe.error(convertToBaseException(throwable));
                            }
                        });
            }
            if (isCompletable) {
                return ((Completable) wrapped.adapt(call)).onErrorResumeNext(
                        new Function<Throwable, CompletableSource>() {
                            @Override
                            public CompletableSource apply(Throwable throwable) throws Exception {
                                return Completable.error(convertToBaseException(throwable));
                            }
                        });
            }
            return ((Observable) wrapped.adapt(call)).onErrorResumeNext(
                    new Function<Throwable, ObservableSource>() {
                        @Override
                        public ObservableSource apply(Throwable throwable) throws Exception {
                            return Observable.error(convertToBaseException(throwable));
                        }
                    });
        }

        private BaseException convertToBaseException(Throwable throwable) {
            if (throwable instanceof BaseException) {
                return (BaseException) throwable;
            }

            if (throwable instanceof IOException) {
                return BaseException.toNetworkError(throwable);
            }

            if (throwable instanceof HttpException) {
                HttpException httpException = (HttpException) throwable;
                Response response = httpException.response();
                if (response.errorBody() != null) {
                    try {
                        ErrorResponse errorResponse =
                                new Gson().fromJson(response.errorBody().string(),
                                        ErrorResponse.class);
                        if (errorResponse != null && !TextUtils.isEmpty(
                                errorResponse.getMessage())) {
                            return BaseException.toServerError(errorResponse);
                        } else {
                            return BaseException.toHttpError(response);
                        }
                    } catch (IOException e) {
                        Log.e(TAG, e.getMessage());
                        return BaseException.toUnexpectedError(throwable);
                    }
                } else {
                    return BaseException.toHttpError(response);
                }
            }

            return BaseException.toUnexpectedError(throwable);
        }
    }
}

File: app/src/main/java/com/ccc/nameapp/data/source/remote/api/middleware/RxErrorHandlingCallAdapterFactory.kt
package com.ccc.nameapp.data.source.remote.api.middleware

import android.text.TextUtils
import android.util.Log
import com.ccc.nameapp.data.source.remote.api.error.BaseException
import com.ccc.nameapp.data.source.remote.api.error.ErrorResponse
import com.google.gson.Gson
import io.reactivex.*
import retrofit2.Call
import retrofit2.CallAdapter
import retrofit2.HttpException
import retrofit2.Retrofit
import retrofit2.adapter.rxjava2.RxJava2CallAdapterFactory
import java.io.IOException
import java.lang.reflect.ParameterizedType
import java.lang.reflect.Type

/**
 * ErrorHandling:
 * http://bytes.babbel.com/en/articles/2016-03-16-retrofit2-rxjava-error-handling.html
 * This class only for Call in retrofit 2
 */
class RxErrorHandlingCallAdapterFactory private constructor() : CallAdapter.Factory() {

    private val original: RxJava2CallAdapterFactory = RxJava2CallAdapterFactory.create()

    override fun get(
        returnType: Type,
        annotations: Array<Annotation>,
        retrofit: Retrofit
    ): CallAdapter<*, *> {
        return RxCallAdapterWrapper(
            returnType,
            wrapped = original.get(returnType, annotations, retrofit) as CallAdapter<Any, Any>
        )
    }

    /**
     * RxCallAdapterWrapper
     */
    internal inner class RxCallAdapterWrapper(
        private val returnType: Type,
        private val wrapped: CallAdapter<Any, Any>?
    ) : CallAdapter<Any, Any> {

        override fun responseType(): Type? {
            return wrapped?.responseType()
        }

        override fun adapt(call: Call<Any>): Any? {
            val rawType = getRawType(returnType)

            val isFlowable = rawType == Flowable::class.java
            val isSingle = rawType == Single::class.java
            val isMaybe = rawType == Maybe::class.java
            val isCompletable = rawType == Completable::class.java
            val isInvalidObject = rawType != Observable::class.java && !isFlowable && !isSingle && !isMaybe
            if (isInvalidObject) {
                return null
            }
            if (returnType !is ParameterizedType) {
                val name = when {
                    isFlowable -> "Flowable"
                    isSingle -> "Single"
                    isMaybe -> "Maybe"
                    else -> "Observable"
                }
                val errorMessage = "$name return type must be parameterized as $name <Foo> or $name <? extends Foo>"
                throw IllegalStateException(errorMessage)
            }

            return when {
                isFlowable -> (wrapped?.adapt(call) as Flowable<*>).onErrorResumeNext { throwable: Throwable ->
                    Flowable.error(
                        convertToBaseException(throwable)
                    )
                }
                isSingle -> (wrapped?.adapt(call) as Single<*>).onErrorResumeNext { throwable ->
                    Single.error(
                        convertToBaseException(throwable)
                    )
                }
                isMaybe -> (wrapped?.adapt(call) as Maybe<*>).onErrorResumeNext { throwable: Throwable ->
                    Maybe.error(
                        convertToBaseException(throwable)
                    )
                }
                isCompletable -> (wrapped?.adapt(call) as Completable).onErrorResumeNext { throwable ->
                    Completable.error(
                        convertToBaseException(throwable)
                    )
                }
                else -> (wrapped?.adapt(call) as Observable<*>).onErrorResumeNext { throwable: Throwable ->
                    Observable.error(
                        convertToBaseException(throwable)
                    )
                }
            }
        }

        private fun convertToBaseException(throwable: Throwable): BaseException {
            if (throwable is BaseException) {
                return throwable
            }

            if (throwable is IOException) {
                return BaseException.toNetworkError(throwable)
            }

            if (throwable is HttpException) {
                val response = throwable.response()
                if (response.code() in 500..511) {
                    return try {
                        val errorResponse = Gson().fromJson(
                            response.errorBody()?.string(),
                            ErrorResponse::class.java
                        )
                        if (errorResponse != null && !TextUtils.isEmpty(errorResponse.message)) {
                            BaseException.toServerError(errorResponse)
                        } else {
                            BaseException.toHttpError(response)
                        }
                    } catch (e: IOException) {
                        Log.e(TAG, e.message ?: "null")
                        BaseException.toUnexpectedError(throwable)
                    }
                } else {
                    return BaseException.toHttpError(response)
                }
            }

            return BaseException.toUnexpectedError(throwable)
        }
    }

    companion object {

        private val TAG = RxErrorHandlingCallAdapterFactory::class.java.name

        fun create(): CallAdapter.Factory {
            return RxErrorHandlingCallAdapterFactory()
        }
    }
}
